lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp

```diff
@@ -635,12 +635,14 @@ class BroadcastOpConversionPattern
     mlir::Value input = srcOp.getOperand(0);
     mlir::Value result = srcOp.getResult();
 
-    if (srcOp->getUsers().empty()) {
-      return rewriter.notifyMatchFailure(
-          srcOp, "ttir.broadcast op should have at least one use.");
+    SmallVector<Operation *> srcOpUsers;
+    for (Operation *nextOp : srcOp->getUsers()) {
+      srcOpUsers.push_back(nextOp);
+    }
+    for (Operation *nextOp : srcOpUsers) {
+      nextOp->replaceUsesOfWith(result, input);
```

Comment:
Can't this be done in a single loop?

Reply:
I found that the Users list of the op is mutable, so I can't modify it while iterating, so the first loop makes copy of the current users and the second loop replaces each of them, so we can't combine them.

Reply:
I don't know if changing user op `nextOp` actually would modify list in the `srcOp`. You can try and see, but anyway, there is a `replaceAllUsesWith` method. Could you avoid both loops with using this?

```
  /// Replace all uses of results of this operation with the provided 'values'.
  template <typename ValuesT>
  void replaceAllUsesWith(ValuesT &&values) {
```

Reply:
If you notice the lines below, that was the original original solution to use rewriter.replaceAllUsesWith(), but somehow it fails for the a test case where is has two back-to-back broadcast ops, both of whom are folded. 

So I am modifying replaceAllUsesWith to use the above loops, as it is probably the other way around that using the above loops actually update modify list in the `srcOp` to correct users and so the test is passing.

Reply:
This is such a weird issue. I've added `target.addLegalOp<ttir::BroadcastOp>();` in `TTIRToTTNNPass.cpp` in order to see which broadcast op remains, and they somehow both remain. I am banging my head since last night to think of why it would be the case, only thing that makes sense is that rewriter is somehow doing things in parallel and it doesn't allow editing op during one matchAndRewrite which is later being removed in another matchAndRewrite, though why it doesn't throw anything is a mystery to me. 

@uazizTT in your fix you are modifying OP in place which throws this assertion for me:
```
Assertion `(replacedRoot() || updatedRootInPlace()) && "expected pattern to replace the root operation"' failed
```

When I change the line 643 to `rewriter.modifyOpInPlace(nextOp, [&]() { nextOp->replaceUsesOfWith(result, input); });` then the fix doesn't work, broadcast ops remain. Which makes sense Looking at the `replaceAllUsesWith` implementation in `/opt/ttmlir-toolchain/include/mlir/IR/PatternMatch.h`, it is doing the same thing as you are manually doing in this PR.

I've tried to use `rewriter.replaceOp(srcOp, result);` instead of:
```
rewriter.replaceAllUsesWith(result, input);
rewriter.eraseOp(srcOp);
```

because it is doing both those things inside, and first time I built it didn't work, broadcast ops remained. Then later when I rebuilt now I get infinite loop when I run it. It is so weird and my best guess is that it is not deterministic because of some rewriter parallelism. I plan to read up on internals of MLIR pattern matching implementation to get some better understanding of this...

Reply:
Woah! looks like I opened pandora box here, yeah there are some rules what is allowed to modify and what not in these pattern matchers. But yeah I am not sure why `replaceAllUsesWith` wouldn't work. 
@mrakitaTT if you have branch with these repros please share, maybe I'll want to play with APIs.

Reply:
@mrakitaTT I tried the test case you sent me and it doesn't throw any assertion errors for me i.e. there is no error due to modifying ops in place. The case also has back-to-back broadcast ops and this method seems to replace them.

I agree that it is a weird problem, but I think the reason this solution works is because I first make list of all the Ops that needs replacement and then commit to replace them, so the modify list is not changed.

Reply:
@rpavlovicTT sure, I am currently on branch `mrakita/issue_805` and here is the test mlir:
```
module attributes {} {
  func.func public @main(%arg0: tensor<32x32xf32> {mhlo.layout_mode = "default", mhlo.sharding = "{replicated}"}) -> (tensor<32x32xf32> {jax.result_info = "", mhlo.layout_mode = "default"}) {
    %0 = "ttir.constant"() <{value = dense<0xFF800000> : tensor<1xf32>}> : () -> tensor<1xf32>
    %1 = tensor.empty() : tensor<32x1xf32>
    %2 = "ttir.max"(%arg0, %1) <{dim_arg = [1 : i32], keep_dim = true, operand_constraints = [#tt.operand_constraint<dram|l1|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any_device_tile>, #tt.operand_constraint<dram|l1|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any_device_tile>]}> : (tensor<32x32xf32>, tensor<32x1xf32>) -> tensor<32x1xf32>
    %3 = tensor.empty() : tensor<32xf32>
    %4 = "ttir.reshape"(%2, %3) <{operand_constraints = [#tt.operand_constraint<dram|l1|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any_device_tile>, #tt.operand_constraint<dram|l1|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any_device_tile>], shape = [32 : i32]}> : (tensor<32x1xf32>, tensor<32xf32>) -> tensor<32xf32>
    %5 = "ttir.constant"() <{value = dense<0xFF800000> : tensor<1xf32>}> : () -> tensor<1xf32>
    %6 = tensor.empty() : tensor<32xf32>
    %7 = "ttir.broadcast"(%5, %6) <{dimension = [], operand_constraints = [#tt.operand_constraint<dram|l1|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any_device_tile>, #tt.operand_constraint<dram|l1|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any_device_tile>]}> : (tensor<1xf32>, tensor<32xf32>) -> tensor<32xf32>
    %8 = tensor.empty() : tensor<32xf32>
    %9 = "ttir.maximum"(%7, %4, %8) <{operandSegmentSizes = array<i32: 2, 1>, operand_constraints = [#tt.operand_constraint<dram|l1|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any_device_tile>, #tt.operand_constraint<dram|l1|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any_device_tile>, #tt.operand_constraint<dram|l1|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any_device_tile>]}> : (tensor<32xf32>, tensor<32xf32>, tensor<32xf32>) -> tensor<32xf32>
    %10 = tensor.empty() : tensor<32x1xf32>
    %11 = "ttir.broadcast"(%9, %10) <{dimension = [0], operand_constraints = [#tt.operand_constraint<dram|l1|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any_device_tile>, #tt.operand_constraint<dram|l1|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any_device_tile>]}> : (tensor<32xf32>, tensor<32x1xf32>) -> tensor<32x1xf32>
    %12 = tensor.empty() : tensor<32x32xf32>
    %13 = "ttir.broadcast"(%11, %12) <{dimension = [0, 1], operand_constraints = [#tt.operand_constraint<dram|l1|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any_device_tile>, #tt.operand_constraint<dram|l1|tile|none|interleaved|single_bank|height_sharded|width_sharded|block_sharded|any_layout|any_device_tile>]}> : (tensor<32x1xf32>, tensor<32x32xf32>) -> tensor<32x32xf32>

    return %13 : tensor<32x32xf32>
  }
}
```

I am using the command `build/bin/ttmlir-opt --ttir-to-ttnn-backend-pipeline test/ttmlir/Conversion/StableHLOToTTIR/test.mlir` to run it.

@uazizTT Let's see what happens when you change the line 643 to `rewriter.modifyOpInPlace(nextOp, [&]() { nextOp->replaceUsesOfWith(result, input); });` (which is what you need to do anyways). If it still passes for you then I'll try on main, I hope there is not some non-determinism happening with different builds.

Reply:
The rule is, if you're inside of a pattern rewriter you **must** use `rewriter` to do any graph modifications.

[replaceAllOpUsesWith](https://mlir.llvm.org/doxygen/classmlir_1_1RewriterBase.html#af11a67366e4e1531f3bb0fbe6a67ec03)
```c++
rewriter.replaceAllOpUsesWith(nextOp, input);
```

Reply:
[Restrictions ¶](https://mlir.llvm.org/docs/PatternRewriter/#restrictions)
> Within the match section of a pattern, the following constraints apply:
>  - No mutation of the IR is allowed.
>
> Within the rewrite section of a pattern, the following constraints apply:
>  - All IR mutations, including creation, must be performed by the given PatternRewriter. This class provides >  hooks for performing all of the possible mutations that may take place within a pattern. For example, this >  means that an operation should not be erased via its erase method. To erase an operation, the appropriate >  PatternRewriter hook (in this case eraseOp) should be used instead.
>  - The root operation is required to either be: updated in-place, replaced, or erased.